### PR TITLE
Adds support to message crossposting

### PR DESCRIPTION
**Pull request by Zaitara**
This PR adds support to crosspost message endpoint

See [relevant discord documentation](https://discord.com/developers/docs/resources/channel#crosspost-message)

This has been tested by crossposting a message

### CODE REVIEW

File: src/model/channel/channel_id.rs [hunk 1]
```diff
@@ -548,6 +548,25 @@ impl ChannelId {
...
+    ///
+    /// Requires either to be the message author or to have manage [Manage Messages] permissions on this channel
+    ///
+    /// **Note**: Only available on announcements channels
+    ///
+    /// # Errors
+    // Returns [`Error::Http`] if the current user lacks permission,
+    // and if the user is not the author of the message
+    /// [Manage Messages]: Permissions::MANAGE_MESSAGES
+
```
Comment (arqunis) [inline src/model/channel/channel_id.rs]:
```suggestion
```

File: src/model/channel/message.rs [hunk 2]
```diff
@@ -130,6 +130,38 @@ pub struct Message {
...
+    ///
+    /// **Note**: Only available on announcements channels
+    /// # Errors
+    ///
+    /// If the `cache` is enabled, returns a
+    /// [`ModelError::InvalidPermissions`] if the current user does not have
+    /// the required permissions.
+    ///
+    /// [Manage Messages]: Permissions::MANAGE_MESSAGES
+
```
Comment (arqunis) [inline src/model/channel/message.rs]:
```suggestion
```

**Review by arqunis — CHANGES_REQUESTED** [review]

File: src/http/client.rs [hunk 3]
```diff
@@ -1033,6 +1033,18 @@ impl Http {
         .await
     }
 
+    pub async fn crosspost_message(&self, channel_id: u64, message_id: u64) -> Result<Message> {
```
Comment (Lakelezz) [inline src/http/client.rs:1039]: Please document this method!

File: src/http/routing.rs [hunk 4]
```diff
@@ -92,6 +92,12 @@ pub enum Route {
     ///
     /// [`ChannelId`]: crate::model::id::ChannelId
     ChannelsIdPinsMessageId(u64),
+    /// Route for the `/channels/:channel_id/message/:message_id/crosspost` path
```
Comment (Lakelezz) [inline src/http/routing.rs]:
```suggestion
    /// Route for the `/channels/:channel_id/message/:message_id/crosspost` path.
```

File: src/http/routing.rs [hunk 5]
```diff
@@ -92,6 +92,12 @@ pub enum Route {
...
+    ///
+    /// The data is the relevant [`ChannelId`]
```
Comment (Lakelezz) [inline src/http/routing.rs]:
```suggestion
    /// The data is the relevant [`ChannelId`].
```

File: src/http/routing.rs [hunk 6]
```diff
@@ -345,6 +351,10 @@ impl Route {
         format!(api!("/channels/{}/messages/{}"), channel_id, message_id)
     }
 
+    pub fn channel_message_crosspost(channel_id: u64, message_id: u64) -> String {
```
Comment (Lakelezz) [inline src/http/routing.rs:354]: Please document this function.

File: src/model/channel/channel_id.rs [hunk 7]
```diff
@@ -548,6 +548,24 @@ impl ChannelId {
         http.as_ref().pin_message(self.0, message_id.into().0).await
     }
 
+    /// Crossposts a [`Message`]
```
Comment (Lakelezz) [inline src/model/channel/channel_id.rs]:
```suggestion
    /// Crossposts a [`Message`].
```

File: src/model/channel/channel_id.rs [hunk 8]
```diff
@@ -548,6 +548,24 @@ impl ChannelId {
...
+    ///
+    /// Requires either to be the message author or to have manage [Manage Messages] permissions on this channel
```
Comment (Lakelezz) [inline src/model/channel/channel_id.rs]:
```suggestion
    /// Requires either to be the message author or to have manage [Manage Messages] permissions on this channel.
```

File: src/model/channel/channel_id.rs [hunk 9]
```diff
@@ -548,6 +548,24 @@ impl ChannelId {
...
+    ///
+    /// **Note**: Only available on announcements channels
```
Comment (Lakelezz) [inline src/model/channel/channel_id.rs]:
```suggestion
    /// **Note**: Only available on announcements channels.
```

File: src/model/channel/channel_id.rs [hunk 10]
```diff
@@ -548,6 +548,24 @@ impl ChannelId {
...
+    ///
+    /// # Errors
+    // Returns [`Error::Http`] if the current user lacks permission,
+    // and if the user is not the author of the message
```
Comment (Lakelezz) [inline src/model/channel/channel_id.rs]:
```suggestion
    // and if the user is not the author of the message.
```

File: src/model/channel/message.rs [hunk 11]
```diff
@@ -130,6 +130,37 @@ pub struct Message {
 
 #[cfg(feature = "model")]
 impl Message {
+    /// Crossposts this message
```
Comment (Lakelezz) [inline src/model/channel/message.rs]:
```suggestion
    /// Crossposts this message.
```

File: src/model/channel/message.rs [hunk 12]
```diff
@@ -130,6 +130,37 @@ pub struct Message {
...
+    ///
+    /// Requires either to be the message author or to have manage [Manage Messages] permissions on this channel
```
Comment (Lakelezz) [inline src/model/channel/message.rs]:
```suggestion
    /// Requires either to be the message author or to have manage [Manage Messages] permissions on this channel.
```

File: src/model/channel/message.rs [hunk 13]
```diff
@@ -130,6 +130,37 @@ pub struct Message {
...
+    ///
+    /// **Note**: Only available on announcements channels
```
Comment (Lakelezz) [inline src/model/channel/message.rs]:
```suggestion
    /// **Note**: Only available on announcements channels.
    ///
```

**Review by Lakelezz — CHANGES_REQUESTED** [review]
Thanks for the pull request!
I added missing full-stops and please document the items I marked too!

File: src/http/routing.rs (hunk 6 above)
Comment (Zaitara) [reply src/http/routing.rs:354]: Others functions here are undocumented, I don't think this one should be documented

**Review by arqunis — APPROVED** [review]

File: src/model/error.rs [hunk 14]
```diff
@@ -184,6 +196,8 @@ impl Display for Error {
             Error::ItemMissing => f.write_str("The required item is missing from the cache."),
             Error::WrongGuild => f.write_str("Provided member or channel is from the wrong guild."),
             Error::MessageTooLong(_) => f.write_str("Message too large."),
+            Error::MessageAlreadyCrossposted => f.write_str("Message already crossposted"),
```
Comment (Lakelezz) [inline src/model/error.rs]:
```suggestion
            Error::MessageAlreadyCrossposted => f.write_str("Message already crossposted."),
```

File: src/model/error.rs [hunk 15]
```diff
@@ -184,6 +196,8 @@ impl Display for Error {
...
+            Error::CannotCrosspostMessage => f.write_str("Cannot crosspost this message type"),
```
Comment (Lakelezz) [inline src/model/error.rs]:
```suggestion
            Error::CannotCrosspostMessage => f.write_str("Cannot crosspost this message type."),
```

**Review by Lakelezz — CHANGES_REQUESTED** [review]
Thanks for fixing my remarks, please add fullstops for the newly added error text.

**Comment by Lakelezz** [comment]
Thanks for your effort! :)

**Pull request merged**